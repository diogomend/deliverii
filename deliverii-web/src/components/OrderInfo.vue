<template>
  <v-card-text>
    <div class="mb-5">
      <div class="heading">Date</div>
      <div class="description">{{ getFormattedDate(created) }}</div>
    </div>
    <div class="mb-5">
      <div class="heading">Delivery Address</div>
      <div class="description">
            <p>{{ address.addr1 }}</p>
            <p>{{ address.addr2 }}</p>
            <p>{{ address.postCode }}, {{ address.city }}</p>
      </div>
    </div>
    <div class="mb-5">
      <div class="heading">Status</div>
      <div class="description">
        <v-btn disabled outlined>{{ status }}</v-btn>
        <template v-if="nextState">
          <v-btn
            :disabled="!isNextStateEnabled"
            color="primary"
            class="ml-5"
            @click="updateToNextStatus"
          >
            {{ nextState.status }}
          </v-btn>
        </template>
      </div>
    </div>
    <div class="mb-5">
      <div class="heading">History</div>
      <div class="description">
          <div v-for="(historySt, i) in history" :key="i">
              {{ getFormattedDate(historySt.date) }}: {{historySt.status}}
          </div>
      </div>
    </div>
    <div v-if="canCancel" class="mt-10">
        <v-btn color="error" bottom right @click="updateOrderState('Canceled')">
            Cancel
        </v-btn>
    </div>
  </v-card-text>
</template>
<script>
import { mapGetters, mapActions } from "vuex";
import { formatDate } from "@/plugins/dayjs";
import { GET_NEXT_STATE } from "@/config/config";
export default {
  props: {
    orderID: {
      required: true
    },
    address: {
      required: true
    },
    status: {
      required: true
    },
    history: {
      required: true
    },
    created: {
      required: true
    },
    customer: {
      required: false
    }
  },
  methods: {
    ...mapActions("order", ["updateStatus"]),
    getFormattedDate(date) {
      return formatDate(date);
    },
    async updateToNextStatus() {
        const { status } = this.nextState;
        await this.updateOrderState(status);
    },
    async updateOrderState(status) {
      const ret = await this.updateStatus({
        orderID: this.orderID,
        status: status
      });
      if (ret) {
        this.$emit("statusChanged", status);
      }
    }
  },
  computed: {
    ...mapGetters({
      user: "auth/getUser"
    }),
    nextState() {
      return GET_NEXT_STATE(this.status);
    },
    canCancel() {
        return !this.user.isManager && this.status == "Placed";
    },
    isNextStateEnabled() {
      const nextState = this.nextState;

      return (
        (nextState.user === "user" && !this.user.isManager) ||
        (nextState.user === "manager" && this.user.isManager)
      );
    }
  }
};
</script>
<style scoped>
.heading {
  font-size: 16px;
  font-weight: bold;
}

.v-application p {
      margin-bottom: 0;
}
</style>
