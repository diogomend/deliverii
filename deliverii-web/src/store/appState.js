const getTimeout = alert => {
  return "timeout" in alert ? alert.timeout : 5000;
};

const actions = {
  toggleDrawer: ({ commit }, toggle) => {
    commit("TOGGLE_DRAWER", toggle);
  },
  toggleLoading: ({ commit }, toggle) => {
    commit("TOGGLE_LOADING", toggle);
  },
  addAlert: ({ commit, getters }, alert) => {
    const timeout = getTimeout(alert);
    const alerts = getters.getAlerts;
    const repeatedIndex = alerts.findIndex(val => {
      return val.type === alert.type && val.message === alert.message;
    });
    let id;

    do {
      id = Date.now() + Math.random();
    } while (getters.getAlerts.some(elem => elem.id === id));

    commit("ADD_ALERT", { ...alert, id });

    if (repeatedIndex !== -1) {
      commit("REMOVE_ALERT", repeatedIndex);
    }

    setTimeout(() => {
      const index = getters.getAlerts.findIndex(elem => elem.id === id);

      if (index !== -1) {
        commit("REMOVE_ALERT", index);
      }
    }, timeout);
  }
};

const mutations = {
  ADD_ALERT(state, { type, message, id }) {
    state.alert.push({
      type,
      message,
      id
    });
  },
  REMOVE_ALERT(state, index) {
    state.alert.splice(index, 1);
  },
  TOGGLE_LOADING(state, loading) {
    state.isLoadingSession = loading;
  },
  TOGGLE_DRAWER(state, drawer) {
    state.drawer = drawer;
  }
};

const getters = {
  getAlerts: state => state.alert,
  getLoadingSession: state => state.isLoadingSession,
  drawer: state => state.drawer
};

const state = () => ({
  drawer: false,
  isLoadingSession: true,
  alert: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
