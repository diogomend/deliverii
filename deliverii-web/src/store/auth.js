import auth from "../service/auth";
const actions = {
  login: async ({ commit, dispatch }, { email, password }) => {
    const res = await auth.login(email, password);
    if (res) {
      commit("SET_USER", res.user);
      commit("SET_TOKEN", res.token);
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );
      return res;
    }
    return false;
  },
  register: async ({ dispatch }, registerObj)=> {
    const res = await auth.register(registerObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Registered Successfully`
        },
        { root: true }
      );
      return res;
    }
    return false;
  },
  logout({ commit }) {
    commit("SET_USER", false);
    commit("SET_TOKEN", false);

    auth.logoutUser();
  },

  async loadSession({ commit }) {
    const isLoggedIn = auth.isLoggedIn();
    if (isLoggedIn) {
      const user = auth.getUserInfo();
      commit("SET_USER", user);
    }
  },
};

const mutations = {
  SET_USER(state, user) {
    state.user = user;
  },
  SET_TOKEN(state, token) {
    state.token = token;
  }
};

const getters = {
  getUser: state => state.user,
  isLoggedIn: state => !!state.user
};

const state = () => ({
  user: false,
  token: false
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
