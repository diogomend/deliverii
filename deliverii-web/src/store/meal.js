import meal from "../service/meal";

const actions = {
  fetchMeals: async (_, restaurantID) => {
    return await meal.fetchMeals(restaurantID);
  },
  fetchMeal: async (_, { restaurant, meal: mealID }) => {
    const mealList = await meal.fetchMeal(restaurant, mealID);
    if (mealList && mealList.length) {
      return mealList[0];
    }

    return false;
  },

  updateMeal: async ({ dispatch }, { restaurant, mealID, mealObj }) => {
    const res = await meal.updateMeal(restaurant, mealID, mealObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );

      return true;
    }

    return false;
  },

  createMeal: async ({ dispatch }, { restaurant, mealObj }) => {
    const res = await meal.createMeal(restaurant, mealObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Success`
        },
        { root: true }
      );

      return true;
    }

    return false;
  }
};

const mutations = {};

const getters = {};

const state = () => ({
  meals: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
