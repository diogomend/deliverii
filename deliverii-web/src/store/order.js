import order from "../service/order";

const actions = {
  fetchAll: async ({ commit }, id) => {
    const res = await order.fetchAll(id);
    commit("ADD_ORDERS", res);
  },
  fetchOrder: async (_, id) => {
    return await order.fetchOrder(id);
  },
  updateStatus: async (_, { orderID, status }) => {
    return await order.updateStatus(orderID, status);
  },
  loadCart: async ({ commit }) => {
    const cart = await order.loadCart();
    if (cart) {
      commit("ADD_CART", JSON.parse(cart));
    }
  },
  addCart({ commit }, cart) {
    order.saveCart(cart);
    commit("ADD_CART", cart);
  },
  addAddress({ commit }, address) {
    commit("ADD_ADDRESS", address);
  },
  async makeOrder({ dispatch, getters }) {
    const cartObj = getters.getCart;
    const meals = cartObj.cart.map(item => {
      return {
        meal: item._id,
        quantity: item.quantity
      };
    });

    const request = {
      meals: meals,
      restaurantID: cartObj.restaurantID,
      address: cartObj.address
    };

    const ret = await order.makeOrder(request);
    if (ret._id) {
      await dispatch(
        "appState/addAlert",
        {
          type: "success",
          message: `Order created successfully`
        },
        { root: true }
      );
      return ret._id;
    }

    return false;
  }
};

const mutations = {
  ADD_ORDERS(state, orders) {
    state.orders = orders;
  },
  ADD_CART(state, cart) {
    state.cart = cart;
  },
  ADD_ADDRESS(state, address) {
    state.cart.address = address;
  }
};

const getters = {
  getOrders: state => state.orders,
  getCart: state => state.cart
};

const state = () => ({
  orders: [],
  cart: {}
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
