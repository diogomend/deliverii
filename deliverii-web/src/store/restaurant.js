import restaurant from "../service/restaurant";

const actions = {
  fetchAll: async ({ commit }) => {
    const res = await restaurant.fetchAll();
    commit("ADD_RESTAURANTS", res);
  },
  fetchRestaurantInfo: async (_, id) => {
    return await restaurant.fetchRestaurant(id);
  },
  deleteRestaurant: async ({ dispatch }, id) => {
    const res = await restaurant.deleteRestaurant(id);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Delete successfully` },
        { root: true }
      );
    }

    return res;
  },
  createRestaurant: async ({ dispatch }, restaurantObj) => {
    const res = await restaurant.createRestaurant(restaurantObj);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Created successfully` },
        { root: true }
      );
    }

    return res;
  },

  addBlacklist: async ({ dispatch }, { restaurantID, customerID }) => {
    const res = await restaurant.addBlacklist(restaurantID, customerID);
    if (res) {
      await dispatch(
        "appState/addAlert",
        { type: "success", message: `Added to blacklist` },
        { root: true }
      );
    }

    return res;
  }
};

const mutations = {
  ADD_RESTAURANTS(state, restaurants) {
    state.restaurants = restaurants;
  }
};

const getters = {
  getRestaurants: state => state.restaurants
};

const state = () => ({
  restaurants: []
});

export default {
  namespaced: true,
  getters,
  actions,
  state,
  mutations
};
