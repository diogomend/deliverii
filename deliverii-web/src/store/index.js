import Vue from "vue";
import Vuex from "vuex";
import appState from "@/store/appState";
import auth from "@/store/auth";
import order from "@/store/order";
import restaurant from "@/store/restaurant";
import meal from "@/store/meal";

Vue.use(Vuex);

export const storeOptions = {
  modules: {
    appState,
    auth,
    order,
    restaurant,
    meal
  }
};

export default new Vuex.Store(storeOptions);
