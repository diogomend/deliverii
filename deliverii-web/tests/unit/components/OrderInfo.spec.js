import Vuetify from 'vuetify';
import Vue from 'vue';
import Vuex from "vuex";
import OrderInfo from "@/components/OrderInfo.vue";
import { shallowMount } from "@vue/test-utils"
import { storeOptions } from "@/store/index"

Vue.use(Vuex);
Vue.use(Vuetify);


describe("OrderInfo.vue", () => {
    let store;

    beforeEach(() => {
        jest.clearAllMocks();
        store = new Vuex.Store(storeOptions);
      });

      it("is a Vue instance", () => {
        const wrapper = shallowMount(OrderInfo, {
          store,
          propsData: {
              address: {},
              created: '2019-11-10'
          }
        });
    
        const template = wrapper.html();
        expect(template).toMatchSnapshot();
      });
})