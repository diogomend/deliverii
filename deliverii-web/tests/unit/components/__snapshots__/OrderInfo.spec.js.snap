// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderInfo.vue is a Vue instance 1`] = `
<v-card-text-stub>
  <div class="mb-5">
    <div class="heading">Date</div>
    <div class="description">2019-11-10 00:00</div>
  </div>
  <div class="mb-5">
    <div class="heading">Delivery Address</div>
    <div class="description">
      <p></p>
      <p></p>
      <p>, </p>
    </div>
  </div>
  <div class="mb-5">
    <div class="heading">Status</div>
    <div class="description">
      <v-btn-stub tag="button" activeclass="" disabled="true" outlined="true" type="button"></v-btn-stub>
      <!---->
    </div>
  </div>
  <div class="mb-5">
    <div class="heading">History</div>
    <div class="description"></div>
  </div>
  <!---->
</v-card-text-stub>
`;
